src/librustc_typeck/check/demand.rs

```diff
@@ -239,6 +239,19 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                 }
                 None
             }
+            (expected_ty, &ty::TyRef(_, checked_ty)) => {
+                if expected_ty == &checked_ty.ty.sty {
```

Comment:
I think it would be cool if this would infinitely deref, so we'd get suggestions for `&&T` when `T` was expected, or even `&str` when starting out with a `String`.